data_layout.go

```diff
@@ -0,0 +1,421 @@
+	mangling                  ManglingStyle                          // depends on os - "e" for linux, "o" for mac
+	nativeIntBitWidths        []uint64                               // default - depends on arch "8:16:32:64" for x86-64
+	nonIntegralPointerTypes   []string                               // Unsure of the datatype, so leaving it as string to accommodate anything.
+}
+
+// NewDataLayout returns an instance of DataLayout with default values
+// taken from https://llvm.org/docs/LangRef.html#data-layout
+func NewDataLayout(os string, arch string) *DataLayout {
+	dl := &DataLayout{
+		isBigEndian:               true,
+		naturalStackAlignment:     0,
+		programMemoryAddressSpace: 0,
+		globalVarAddressSpace:     0,
+		allocaAddressSpace:        0,
+		ptrSizeAlign:              getDefaultPtrAligns(),
+		intSizeAlign:              getDefaultIntAligns(),
+		vecSizeAlign:              getDefaultVectorAligns(),
+		floatSizeAlign:            getDefaultFloatAligns(),
+		aggAlign:                  NewAggregateAlignment(0, 64),
+	}
+	switch arch {
+	case "x86-64":
+		dl.nativeIntBitWidths = []uint64{8, 16, 32, 64}
+	}
+
+	if os == "linux" {
+		dl.mangling = ELF
+	} else if strings.HasPrefix(os, "darwin") || strings.HasPrefix(os, "macosx") {
+		dl.mangling = MachO
+	}
+	return dl
+}
+
+// NewDataLayoutFromString parses a datalayout string and constructs a DataLayout object.
+// Default values whereever applicable will be added if options are ommitted.
+//
+// This API expects the string to be valid as per LLVM spec, so there aren't many validations.
+func NewDataLayoutFromString(layoutString, os, arch string) *DataLayout {
+	dl := NewDataLayout(os, arch)
+	specs := strings.Split(layoutString, "-")
+	for _, spec := range specs {
+		if spec == "e" {
+			dl.isBigEndian = false
+		} else if spec == "E" {
+			dl.isBigEndian = true
+		} else if strings.HasPrefix(spec, "S") {
+			dl.naturalStackAlignment = getUInt64(strings.TrimPrefix(spec, "S"))
+		} else if strings.HasPrefix(spec, "P") {
+			dl.programMemoryAddressSpace = getUInt64(strings.TrimPrefix(spec, "P"))
+		} else if strings.HasPrefix(spec, "G") {
+			dl.globalVarAddressSpace = getUInt64(strings.TrimPrefix(spec, "G"))
+		} else if strings.HasPrefix(spec, "A") {
+			dl.allocaAddressSpace = getUInt64(strings.TrimPrefix(spec, "A"))
+		} else if strings.HasPrefix(spec, "p") {
+			addPtrSizeAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "i") {
+			addIntSizeAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "v") {
+			addVectorSizeAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "f") {
+			addFloatSizeAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "a") {
+			addAggAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "F") {
+			addFuncPtrAlignFromString(spec, dl)
+		} else if strings.HasPrefix(spec, "m") {
+			dl.mangling = ManglingStyle(strings.Split(spec, ":")[1])
```

Comment:
We should add error detection to prevent the library from crashing if `len(strings.Split(spec, ":"))` is less than 2. In general the library should on panic when we have code paths that are not fully implemented (e.g. a switch missing an enum value case branch). But for code that parses arbitrary user input, we should return an error rather than crashing with a panic.